src/sys/windows/selector.rs

```diff
@@ -166,16 +186,59 @@ impl Registration {
     /// Eventually this function will probably also be modified to handle the
     /// `level()` polling option.
     pub fn push_event(&mut self, set: EventSet, events: &mut Vec<IoEvent>) {
+        trace!("push_event; token={:?}; set={:?}; opts={:?}", self.token, set, self.opts);
+
         // If we're not actually interested in any of these events,
         // discard the event, and then if we're actually delivering an event we
         // stop listening if it's also a oneshot.
         let set = self.interest & set;
+
         if set != EventSet::none() {
-            events.push(IoEvent::new(set, self.token));
+            let event = IoEvent::new(set, self.token);
+
+            if self.opts.is_edge() {
+                events.push(event);
+
+                if self.opts.is_oneshot() {
+                    trace!("deregistering because of oneshot");
+                    self.interest = EventSet::none();
+                }
+            } else {
+                let selector = self.selector.as_ref()
+                    .expect("expected a selector");
+
+                let mut level = selector.level_triggered.lock().unwrap();
+
+                match level.entry(self.key.expect("expected registration key")) {
```

Comment:
You could also do this match via `or_insert(event)` and then or-in the kind, it at least avoids the `Entry` import (but works either way)